src/Common.OData.ApiExplorer/AspNet.OData/DefaultModelTypeBuilder.cs

```diff
@@ -105,8 +105,12 @@ static Type GenerateTypeIfNeeded( IEdmStructuredType structuredType, BuilderCont
             {
                 if ( !structuralProperties.TryGetValue( property.Name, out var structuralProperty ) )
                 {
-                    clrTypeMatchesEdmType = false;
-                    continue;
+                    var name = GetNameFromAttribute( property );
+                    if ( string.IsNullOrEmpty( name ) || !structuralProperties.TryGetValue( name, out structuralProperty ) )
+                    {
+                        clrTypeMatchesEdmType = false;
```

Comment:
This is _close_, but not quite. Let's refactor here and the few lines above to:

```c#
var properties = new List<ClassProperty>();
var structuralProperties = new Dictionary<string, IEdmProperty>( StringComparer.OrdinalIgnoreCase );
var mappedClrProperties = new Dictionary<PropertyInfo, IEdmProperty>();
var clrTypeMatchesEdmType = true;
var hasUnfinishedTypes = false;
var dependentProperties = new List<PropertyDependency>();

foreach ( var property in structuredType.Properties() )
{
    var clrProperty = edmModel.GetAnnotationValue<ClrPropertyInfoAnnotation>( property ).ClrPropertyInfo;

    structuralProperties.Add( property.Name, property );
    mappedClrProperties.Add( clrProperty, property );
}

foreach ( var property in clrType.GetProperties( bindingFlags ) )
{
    if ( !structuralProperties.TryGetValue( property.Name, out var structuralProperty ) &&
         !mappedClrProperties.TryGetValue( property, out structuralProperty ) )
    {
        clrTypeMatchesEdmType = false;
        continue;
    }
```

Reply:
Done. I also added checking for null as `edmModel.GetAnnotationValue<ClrPropertyInfoAnnotation>(property) ` returns null for non-annotated properties.